### PR TITLE
Issue 43349: SampleTypeUpdateServiceDI.getExistingRows() performance

### PR DESCRIPTION
**Pull request by XingY**
#### Rationale
Sample import with merge is slow due to lineage queries (one for parent material and one for parent data) that's done for each row of data in SampleTypeUpdateServiceDI.getExistingRows(). Those queries are needed for sample detailed audit log, which is used by sample timeline. This PR merges data and material parent into one query, and additionally handle 50 rows in a single query, which should reduce the number of lineage queries by ~99%. 

#### Related Pull Requests
* 

#### Changes
* new ExperimentServiceImpl.getParentMaterialAndDataMap function that gets both parent data and parent material for a set of sample lsid in a single query
* SampleTypeUpdateServiceDI.getExistingRows to use getParentMaterialAndDataMap
* Update ExistingDataIteratorsGetRows to get 50 rows a time

### CODE REVIEW

File: experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java [hunk 1]
```diff
@@ -516,10 +521,68 @@ else if (parent instanceof ExpMaterial)
...
+
+    private Map<Integer, Map<String, Object>> getMaterialMapsWithInput(Map<Integer, Map<String, Object>> keys, User user, Container container)
             throws QueryUpdateServiceException
     {
-        return getRows(user, container, keys, true /* this is for detailed audit logging */);
+        Map<Integer, Map<String, Object>> sampleRows = new LinkedHashMap<>();
+        Set<Identifiable> lsids = new LinkedHashSet<>(keys.size());
+        Map<Integer, String> rowLsid = new HashMap<>();
+
+        for (Map.Entry<Integer, Map<String, Object>> keyMap : keys.entrySet())
```
Comment (labkey-matthewb) [inline experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java:540]: The remaining work is to unroll this loop and use using and IN() filter instead of =
(can be separate PR)

**Review by labkey-kevink — APPROVED** [review]
The single lineage query for both parents and children for a group of samples looks good.

File: experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java [hunk 2]
```diff
@@ -516,10 +521,68 @@ else if (parent instanceof ExpMaterial)
...
+                filter = new SimpleFilter(FieldKey.fromParts(ExpMaterialTable.Column.LSID), lsid);
+            else
+                throw new QueryUpdateServiceException("Either RowId or LSID is required to get Sample Type Material.");
+
+            Map<String, Object> sampleRow = new TableSelector(getQueryTable(), filter, null).getMap();
+            if (null != sampleRow)
+            {
+                String sampleLsid = lsid != null ? lsid : (String) sampleRow.get("lsid");
+
+                Identifiable id = LsidManager.get().getObject(sampleLsid);
```
Comment (labkey-matthewb) [inline experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java]: We really shouldn't call LsidManager.get().getObject(sampleLsid);  This causes two more queries per row.  One for findType() and one for getObject().  We already know the type and we already have the row in memory.

**Review by labkey-matthewb — CHANGES_REQUESTED** [review]
I think it would be worth looking at the LsidManager.get().getObject(sampleLsid) issue and deciding whether to change in this PR or fix later.

**Review by labkey-matthewb — APPROVED** [review]
